poetry/utils/pip.py

```diff
@@ -44,20 +45,30 @@ def pip_install(
     args.append(str(path))
 
     try:
-        return environment.run_pip(*args)
+        environment.run_pip(*args)
     except EnvCommandError as e:
         if sys.version_info < (3, 7) and not is_wheel:
             # Under certain Python3.6 installs vendored pip wheel does not contain zip-safe
             # pep517 lib. In this cases we create an isolated ephemeral virtual environment.
             with ephemeral_environment(
                 executable=environment.python, with_pip=True, with_setuptools=True
             ) as env:
-                return environment.run(
+                environment.run(
                     env._bin("pip"),
                     *args,
                     env={**os.environ, "PYTHONPATH": str(env.purelib)},
                 )
+            return 0
+
+        output = decode(e.e.output)
+        if (
+            "KeyboardInterrupt" in output
+            or "ERROR: Operation cancelled by user" in output
+        ):
+            return -2
+
         raise PoetryException(f"Failed to install {path.as_posix()}") from e
+    return 0
```

Comment:
Does this mean we do not propagate the command return code? Additionally, this also means that we loose debug information when pip fails to install things. 

Also, the fix for returning "int" code will be to pass `call=True` to `run_pip` as this will use `subrprocess.call()` insteald of `subprocess.run()`.

Reply:
I was going off of what @sdispater pointed me to in executor.py so it sounds like both need to change then? https://github.com/python-poetry/poetry/blob/f1b1cf9345837b916e15837d0258f1cc6a0b3f91/poetry/installation/executor.py#L324-L337 

I suppose one part of the answer is to just change both these places to say `call=True` - I assume that this will catch keyboard interrupts too

Reply:
Might need to check this. Might also be a reasonably to move logic to `run_pip` or `run` in `poetry.util.env`.

Reply:
While we're at it, I think we aren't supposed to be using `subprocess.call` anyway because we're [on python3.6+](https://github.com/python-poetry/poetry/blob/master/pyproject.toml#L28) and because you [_can_ get the returncode](https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.returncode) from a `subprocess.CompletedProcess` object (which is returned by `subprocess.run()`)
From Python docs:
> The run() function was added in Python 3.5; if you need to retain compatibility with older versions, see the Older high-level API section.

Reply:
@abn No, we don't propagate the return code of the subprocess. The executor only needs to be aware of whether the command succeeded (return code 0) or was interrupted (return code -1 or if the output contains `KeyboardInterrupt` or `ERROR: Operation cancelled by user`). Any other error will be raised and displayed to the user so that does not prevent debugging.

Reply:
Hmm. Won't with this change, the output of the pip install command in the subprocess run be lost? Maybe that's what's happening today anyway and it's a non-issue. I haven't checked if all failure modes (eg: successful completion but build failure) raises a command error. If this is the case, we should not handle this in run_pip anyway and the method should always return an int.

Reply:
> Won't with this change, the output of the pip install command in the subprocess run be lost?

It won't if there is an error since an EnvCommandError will be raised which contains the return code and output. In case of success, it does not matter since we don't need to display it to the end user.

Reply:
I've added tests for this behaviour - a CalledProcessError from subprocess raises an EnvCommandError in poetry (env.run), but a KeyboardInterrupt is just raised as-is and not caught. We'll have to catch it somewhere downstream like in the example from @sdispater. But I think the better way to do it is to look for the return code -2 rather than the string `KeyboardInterrupt` / `Operation cancelled by user`.

This code will also need to be changed, I think, because EnvCommandErrors are a disjoint set from KeyboardInterrupts, so if my understanding is correct, line 226 will never fire because it's impossible for the returncode of an EnvCommandError to be -2.
https://github.com/python-poetry/poetry/blob/master/poetry/installation/executor.py#L222-L228

Reply:
> But I think the better way to do it is to look for the return code -2 rather than the string KeyboardInterrupt / Operation cancelled by user.

That won't work since `pip` might return a return code of `1` in case of `KeyboardInterrupt`. That's why I added this test to ensure we have a `-2` return code.

Reply:
ah! that's annoying :P Alright, I'll keep that check and also add a test for it